### PR TITLE
slight refactor, added auto termination

### PR DESCRIPTION
**Pull request by ashleyschuett**

### CODE REVIEW

**Comment by ashleyschuett** [comment]
depends on https://github.com/containership/containership.cloud.constants/pull/9

File: lib/auto_termination.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
+'use strict';
+
+const cluster_snapshot = require('./cluster_snapshot');
+
+const async = require('async');
+const constants = require('./constants');
```
Comment (normanjoyner) [inline lib/auto_termination.js]: move up

File: lib/auto_termination.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                                }
+
+                                return callback();
+                            });
+                        }
+
+                        return callback();
+                    }
+                }, (err) => {
+                    if(!err) {
```
Comment (normanjoyner) [inline lib/auto_termination.js]: log that we're refusing to terminate the cluster

File: lib/auto_termination.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            timeout: 5000,
+            headers: {
+                Authorization: `Bearer ${termination_options.api_key}`
+            }
+        };
+
+        request(options, (err, response) => {
+            if(err) {
+                return callback(err);
+            } else if(response && response.statusCode !== 201) {
```
Comment (normanjoyner) [inline lib/auto_termination.js]: is `201` the correct response code?

File: lib/cluster_snapshot.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        const options = {
+            url: `${constants.environment.CLOUD_API_BASE_URL}/v2/organizations/${snapshot_options.organization}/backups`,
+            method: 'POST',
+            timeout: 5000,
+            headers: {
+                Authorization: `Bearer ${snapshot_options.api_key}`
+            },
+            json: {
+                cluster_id: snapshot_options.cluster_id,
+                notes: `Scheduled ContainerShip Cloud Snapshot (${new Date().toISOString()})`,
```
Comment (normanjoyner) [inline lib/cluster_snapshot.js]: thoughts on removing this date given we have the created date associated with backup already?

File: lib/leader.js [hunk 5]
```diff
@@ -144,15 +145,33 @@ module.exports = {
...
+                            set_auto_termination_configuration: (callback) => {
+                                const auto_termination_configuration = response.body.auto_termination_configuration;
+
+                                if(!cache.auto_termination_configuration || !_.isEqual(cache.auto_termination_configuration, auto_termination_configuration)) {
+                                    core.cluster.myriad.persistence.set(constants.myriad.AUTO_TERMINATION_CONFIGURATION, JSON.stringify(auto_termination_configuration), (err) => {
+                                        if(err) {
+                                            core.loggers['containership-cloud'].log('warn', `Error persisting auto termination configuration to myriad-kv: ${err.message}`);
+                                        } else {
+                                            cache.auto_termination_configuration = auto_termination_configuration;
+                                            auto_termination.schedule(core, cache, config);
```
Comment (normanjoyner) [inline lib/leader.js]: need to `auto_termination.cancel()` below on non-controlling leaders

File: lib/cluster_snapshot.js (hunk 4 above)
Comment (ashleyschuett) [inline lib/cluster_snapshot.js]: i like it being in the title.. but I also think the user should be able to name the snapshot what they want.

**Comment by nicktate** [comment]
@ashleyschuett pleaes update with `containership.cloud.constants@1.1.9`

**Comment by ashleyschuett** [comment]
@nicktate updated

**Pull request merged**